### ISSUE TITLE
Problem using custom packages in custom defined modules.

### ISSUE THREAD

**Issue by spinus** [NixOS/nixops#434]
I have few custom packages and custom modules.

In nixops I have custom packages defined as:
```
pkgs = (import <nixpkgs> {}) // { my-package = /*...*/ };
```
When I use want to install my custom package defined in nixops config file (for example in a service or systemPackages), that works.
When I want to use custom package in module, that doesn't work (attribute with custom package is not found).

I import modules using ``imports = [ ./path/to/module.nix ];`` syntax.
I think this does not work, because ``pkgs`` is not passed correctly to ``eval-config``, but I'm not sure how to pass it or make it working the other way.

**Comment by ip1981** [comment]
`nixpkgs.config.packageOverrides` should do the trick. Use it in a separate module.

**Comment by spinus** [comment]
@ip1981 I think it's not that. I don't know how to pass modified "pkgs" to imported modules.
So even if I tried what you write, modified version of pkgs is not passed correctly. 
Is there any nixops way to maybe use modified pkgs from my-packages as default "pkgs"?

More or less my network.nix
```
{system}:{

host1 = 
let pkgs = import ./my-packages.nix {inherit system;}
in {
  imports = [
    ./module1.nix # when package is used inside module it does not work
  ];
  environment.systemPackages = [pkgs.my-package]; # <- that works
  services.my-service.enable = true; # <- does not work, because my-package is not available inside my-module.nix
}
};
```

**Comment by spinus** [comment]
Or maybe there is another way of providing custom packages/services in nixops (without forking nixpkgs)?

**Comment by spinus** [comment]
@ip1981 Finally I understand what did you mean and where to put it (at the beginning I've put that on network level rather then host level). Works perfectly. Thank you.

working example would be:
```
{system}: {
  host1 = {
    imports = [ ./module1.nix ];  # now module can use packages overriden/declared below
    nixpkgs.config.overridePackages ....
  }
}
```
Also, nixpkgs can be defined inside a module, so you can group packages in separate module.

I think would be nice to add this information to manual.

**Comment by domenkozar** [comment]
This has nothing to do with NixOps, but rather NixOS which is already documented. I don't think we should duplicate documentation from NixOS since 99% of it applies for NixOps.

Do you have some suggestions for the manual for this to be more clear @spinus ?

**Comment by k0001** [comment]
I've just encountered this situation too, so I would like to ask something related to this: Is there a way to specify the `<nixpkgs>` to be used by the machines from within one of the NixOps files themselves, just like it is possible to specify `nixops.config` as explained here (instead of through `NIXPKGS_CONFIG`)?

**Comment by k0001** [comment]
Also a related question: Shouldn't `NIXPKGS_CONFIG` work for NixOps too, instead of having to go through setting `nixops.config` in the machine definition?

**Comment by spinus** [comment]
@domenkozar take a look on PR, could that be useful?

**Comment by spinus** [comment]
@k0001 yes, in your case if you want to use new namespace or override nixpkgs and use own version/fork you can do that with ```-I``` like ```nixops deploy -I nixpkgs=/path/to/your/nixpkgs``` (or ```nixops deploy -I /path/to/your/nixpkgs```, one of those, don't remember exactly which one.

**Comment by k0001** [comment]
@spinus Thanks. That works for `nixpkgs`, but not for reading `nixpkgs.config` from `NIXPKGS_CONFIG` as it usually happens with various other `nix-*` tools. 

In any case, I worked around this situation by doing this in the physical machine description: 

```nix
nixpkgs.config =
  let x = builtins.getEnv "NIXPKGS_CONFIG";
  in if x == "" then {} else import x;
```

I'm not sure if it is a good idea, but it serves my current needs at least.

**Comment by qknight** [comment]
@spinus 
this does not cover extending `pkgs` (for this you need https://github.com/NixOS/nixpkgs/pull/26321/files)

but if you want to add a new service (which itself brings its own bundled pkgs) you can do this:

https://github.com/nixcloud/rebuild-on-change-daemon
